File: c_files/frequency.c
/**
 * @file frequency.c
 * @brief This program reads a string and prints the frequency of each character in the string.
 * @details The program reads a string and counts the frequency of each character in the string. The program then prints the character and its frequency in ascending order of the character.
 * @author Arthur Clemente Machado
 * @date 2024-05-30
*/
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

// structure to store the character and its frequency
typedef struct data
{
    char c;
    int frequency;
}data;

// function prototypes
void quicksort(data* output, int n);
void quicksort_private(data* output, int esq, int dir);
void insertion(data*output, int n);

/**
 * @brief 
 * 
 * @return int
*/
int main(void){
    int k, frequency;

    data output[1001];
    bool check[1001];
    char input[1001];
    // assume there is a next case before the loop
    bool is_next_case = (scanf(" %[^\r\n]%*c", input) != EOF);

    do
    {    
        k = 0;
        // reset the check array
        memset(check, false, sizeof(check));

        // loop for each character in the input
        for (size_t i = 0; input[i]; i++)
        {
            frequency = 0;
            // check if the character is already counted
            if (!check[i])
            {
                // count the frequency of the character
                for (size_t j = 0; input[j]; j++)
                {
                    if (input[i] == input[j])
                    {
                        check[j] = true;
                        frequency++;
                    }
                    
                }
                // store the character and its frequency
                output[k].c = input[i];
                output[k++].frequency = frequency;
            }
        }
        
        // sort the output array
        quicksort(output, k);
        insertion(output, k);

        // print the output
        for (size_t i = 0; i < k; i++)
        {
            printf("%d %d\n", (int)output[i].c, output[i].frequency);
        }

        // check if there is a next case
        is_next_case = (scanf(" %[^\r\n]%*c", input) != EOF);

        // print a new line if there is a next case
        if (is_next_case)
        {
            printf("\n");
        }
    }while(is_next_case);
    
    return 0;
}

void quicksort(data* output, int n){
    quicksort_private(output, 0, n-1);
}

void quicksort_private(data*output, int esq, int dir){
    int i = esq;
    int j = dir;
    data pivo = output[(esq+dir)/2];

    while (i <= j)
    {
        while (output[i].c > pivo.c)
        {
            i++;
        }
        while (output[j].c < pivo.c)
        {
            j--;
        }
        if (i <= j)
        {
            data tmp = output[i];
            output[i] = output[j];
            output[j] = tmp;
            i++; j--;
        }
    }
    if (i < dir)
    {
        quicksort_private(output, i, dir);
    }
    if (esq < j)
    {
        quicksort_private(output, esq, j);
    }
}

void insertion(data*output, int n){
    for (size_t i = 0; i < n; i++)
    {
        data tmp = output[i];
        int j = i-1;
        while (j >= 0 && output[j].frequency > tmp.frequency)
        {
            output[j+1] = output[j];
            j--;
        }
        output[j+1] = tmp;
    }
}

File: java_files/Frequency.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * Char_frequency
 */
class Char_frequency {
    char c;
    int frequency;

    public Char_frequency(){
        this.c = '\0';
        this.frequency = 0;
    }
}

public class Frequency {
    public static void insertion(int n, Char_frequency[] list){
        for (int i = 1; i < n; i++) {
            Char_frequency tmp = list[i];
            int j = i - 1;
            while (j >= 0 && tmp.frequency < list[j].frequency) {
                list[j + 1] = list[j];
                j--;
            }
            list[j+1] = tmp;
        }
    }

    public static void quicksort(Char_frequency[]list, int n){
        quicksort(list, 0, n-1);
    }

    private static void quicksort(Char_frequency[]list, int esq, int dir){
        int i = esq;
        int j = dir;
        Char_frequency pivo = list[(esq+dir)/2];
        while (i <= j) {
            while (list[i].c > pivo.c) {
                i++;
            }
            while (list[j].c < pivo.c) {
                j--;
            }
            if (i <= j) {
                Char_frequency tmp = list[i];
                list[i] = list[j];
                list[j] = tmp;
                i++; j--;
            }
        }
        if (esq < j) {
            quicksort(list, esq, j);
        }
        if (i < dir) {
            quicksort(list, i, dir);
        }
    }

    public static void main(String[] args) {
        int k, frequency, break_line;
        String input;
        Scanner scanner = new Scanner(System.in);
        Char_frequency[] list = new Char_frequency[1001];
        Boolean[] check = new Boolean[1001];
        while (scanner.hasNextLine()) {
            input = scanner.nextLine();
            k = 0;
            
            Arrays.fill(check, false);
            for (int i = 0; i < input.length(); i++) {
                frequency = 0;
                if (!check[i]) {
                    
                    for (int j = 0; j < input.length(); j++) {
                        if (input.charAt(i) == input.charAt(j)) {
                            check[j] = true;
                            frequency++;
                        }
                    }
                    list[k] = new Char_frequency();
                    list[k].c = input.charAt(i);
                    list[k].frequency = frequency;
                    k++;
                }

            }

            quicksort(list, k);
            insertion(k, list);

            for (int i = 0; i < k; i++) {
                System.out.println((int)list[i].c + " " + list[i].frequency);
            }
            if (scanner.hasNextLine()) {
                System.out.println();

            }
        }
        scanner.close();
    }
}
